crates/revm/src/handler/handle_types/execution.rs

```diff
@@ -3,20 +3,26 @@ use crate::{
     handler::mainnet,
     interpreter::{CallInputs, CreateInputs, SharedMemory},
     primitives::{db::Database, EVMError, Spec},
-    CallFrame, Context, CreateFrame, Frame, FrameOrResult, FrameResult,
+    CallFrame, Context, CreateFrame, Evm, Frame, FrameOrResult, FrameResult,
 };
-use std::{boxed::Box, sync::Arc};
-
 use revm_interpreter::{
-    CallOutcome, CreateOutcome, EOFCreateInput, EOFCreateOutcome, InterpreterResult,
+    CallOutcome, CreateOutcome, EOFCreateInput, EOFCreateOutcome, InterpreterAction,
+    InterpreterResult,
 };
+use std::{boxed::Box, sync::Arc};
 
 /// Handles first frame return handle.
 pub type LastFrameReturnHandle<'a, EXT, DB> = Arc<
     dyn Fn(&mut Context<EXT, DB>, &mut FrameResult) -> Result<(), EVMError<<DB as Database>::Error>>
         + 'a,
 >;
 
+/// Executes a single frame. Errors can be returned in the EVM context.
+/// If `None` is returned, the frame is instead executed normally through the interpreter.
+pub type ExecuteFrameHandle<'a, EXT, DB> = Arc<
+    dyn Fn(&mut Frame, &mut SharedMemory, &mut Evm<'a, EXT, DB>) -> Option<InterpreterAction> + 'a,
```

Comment:
Do you need full `Evm` here? To be consistent it would be better to have `Context` here.

Reply:
`Host` is implemented on `Evm` so yes

Reply:
Yeah I do think it should be moved to `InnerEvmContext`. In recursion it was needed to be on Evm but this is not the case anymore.

Reply:
Moved `Host` impl to `Context` as there is restriction inside Handler. Context is needed for instructions and other handles.